src/coreclr/vm/stubmgr.cpp

```diff
@@ -1714,43 +1743,60 @@ BOOL ILStubManager::TraceManager(Thread *thread,
         target = target >> 1; // call target is encoded as (addr << 1) | 1
 #endif // TARGET_AMD64
 
-        LOG((LF_CORDB, LL_INFO10000, "ILSM::TraceManager: Unmanaged CALLI case 0x%x\n", target));
+        LOG((LF_CORDB, LL_INFO10000, "ILSM::TraceManager: Unmanaged CALLI case 0x%p\n", target));
         trace->InitForUnmanaged(target);
     }
-    else if (pStubMD->IsStructMarshalStub())
+    else if (pStubMD->IsStepThroughStub())
     {
-        // There's no "target" for struct marshalling stubs
-        // so we have nowhere to tell the debugger to move the breakpoint.
-        return FALSE;
+        MethodDesc* pTargetMD = pStubMD->GetILStubResolver()->GetStubTargetMethodDesc();
+        if (pTargetMD == NULL)
+        {
+            LOG((LF_CORDB, LL_INFO1000, "ILSM::TraceManager: Stub has no target to step through to\n"));
+            return FALSE;
+        }
+
+        LOG((LF_CORDB, LL_INFO1000, "ILSM::TraceManager: Step through to target - 0x%p\n", pTargetMD));
+        target = GetStubTarget(pTargetMD);
+        if (target == NULL)
+            return FALSE;
+
+        trace->InitForManaged(target);
     }
-    else
+    else if (pStubMD->HasMDContextArg())
     {
+        LOG((LF_CORDB, LL_INFO1000, "ILSM::TraceManager: Hidden argument is MethodDesc\n"));
+
         // This is either direct forward P/Invoke or a CLR-to-COM call, the argument is MD
         MethodDesc *pMD = (MethodDesc *)arg;
-
         if (pMD->IsNDirect())
         {
             target = (PCODE)((NDirectMethodDesc *)pMD)->GetNativeNDirectTarget();
-            LOG((LF_CORDB, LL_INFO10000, "ILSM::TraceManager: Forward P/Invoke case 0x%x\n", target));
+            LOG((LF_CORDB, LL_INFO10000, "ILSM::TraceManager: Forward P/Invoke case 0x%p\n", target));
             trace->InitForUnmanaged(target);
         }
 #ifdef FEATURE_COMINTEROP
         else
         {
+            LOG((LF_CORDB, LL_INFO1000, "ILSM::TraceManager: Stub is CLR-to-COM\n"));
             _ASSERTE(pMD->IsComPlusCall());
             ComPlusCallMethodDesc *pCMD = (ComPlusCallMethodDesc *)pMD;
             _ASSERTE(!pCMD->IsStatic() && !pCMD->IsCtor() && "Static methods and constructors are not supported for built-in classic COM");
 
             if (pThis != NULL)
             {
                 target = GetCOMTarget(pThis, pCMD->m_pComPlusCallInfo);
-
-                LOG((LF_CORDB, LL_INFO10000, "ILSM::TraceManager: CLR-to-COM case 0x%x\n", target));
+                LOG((LF_CORDB, LL_INFO10000, "ILSM::TraceManager: CLR-to-COM case 0x%p\n", target));
                 trace->InitForUnmanaged(target);
             }
         }
 #endif // FEATURE_COMINTEROP
     }
+    else
+    {
+        LOG((LF_CORDB, LL_INFO1000, "ILSM::TraceManager: No known target, IL Stub is a leaf\n"));
+        // There's no "target" so we have nowhere to tell the debugger to move the breakpoint.
```

Comment:
Do you expect any ILStub cases to come here by design? Ideally we could make this assert so that if someone adds a new ILStub type in the future and forgets to describe it to the debugger the assert would light up.

Reply:
> Do you expect any ILStub cases to come here by design?

Yep. The ArrayOps stubs.

> Ideally we could make this assert so that if someone adds a new ILStub type in the future and forgets to describe it to the debugger the assert would light up.

There is a check already in `DebuggerStepper::TriggerMethodEnter()` for this case. It causes all sorts of issues if it fails, but it does trigger.